kmir/src/kmir/kmir.py

```diff
@@ -66,70 +67,94 @@ def kcfg_explore(self, label: str | None = None) -> Iterator[KCFGExplore]:
         ) as cts:
             yield KCFGExplore(cts, kcfg_semantics=KMIRSemantics())
 
-    def make_init_config(
-        self, parsed_smir: KInner, start_symbol: KInner | str = 'main', sort: str = 'GeneratedTopCell'
-    ) -> KInner:
-        if isinstance(start_symbol, str):
-            start_symbol = token(start_symbol)
-
-        subst = Subst({'$PGM': parsed_smir, '$STARTSYM': start_symbol})
-        init_config = subst.apply(self.definition.init_config(KSort(sort)))
-        return init_config
+    def _make_function_map(self, smir_info: SMIRInfo) -> KInner:
+        parser = Parser(self.definition)
+        parsed_items: dict[KInner, KInner] = {}
+        for item_name, item in smir_info.items.items():
+            if (
+                not item_name in smir_info.function_symbols_reverse
+                and not item['mono_item_kind']['MonoItemFn']['name'] == 'main'
+            ):
+                _LOGGER.warning(f'Item not found in SMIR: {item_name}')
+                continue
+            elif item['mono_item_kind']['MonoItemFn']['name'] == 'main':
+                _LOGGER.warning(f'Hardcoding name "main" for item: {item_name}')
+                item_name = 'main'
+            item_ty = KApply('ty', [token(smir_info.function_symbols_reverse[item_name])])
+            parsed_item = parser.parse_mir_json(item, 'MonoItem')
+            if not parsed_item:
+                raise ValueError(f'Could not parse MonoItemKind: {parsed_item}')
+            parsed_item_kinner, _ = parsed_item
+            if type(parsed_item_kinner) is KApply and parsed_item_kinner.label.name == 'monoItemWrapper':
+                parsed_items[item_ty] = parsed_item_kinner.args[1]
+        return map_of(parsed_items)
+
+    def _make_type_and_adt_maps(self, smir_info: SMIRInfo) -> tuple[KInner, KInner]:
+        parser = Parser(self.definition)
+        types: dict[KInner, KInner] = {}
+        adts: dict[KInner, KInner] = {}
+        for type in smir_info._smir['types']:
+            parse_result = parser.parse_mir_json(type, 'TypeMapping')
+            assert parse_result is not None
+            type_mapping, _ = parse_result
+            assert isinstance(type_mapping, KApply) and len(type_mapping.args) == 2
+            ty, tyinfo = type_mapping.args
+            if ty in types:
+                raise ValueError(f'Key collision in type map: {ty}')
+            types[ty] = tyinfo
+            if isinstance(tyinfo, KApply):
+                if tyinfo.label.name in ['TypeInfo::EnumType', 'TypeInfo::StructType']:
```

Comment:
`StructUnion` also has an `adt_def` field. However, we are not supporting it, other than creating the metadata and parsing it.

Reply:
I guess we can make that change when we have an example for it.